### ISSUE TITLE
in page.module.css, class names with dashes does not appear to work as expected

### ISSUE THREAD

**Issue by pkellner** [vercel/next.js#48759]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 20:59:58 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T6020
    Binaries:
      Node: 18.15.0
      npm: 9.5.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.3.1
      eslint-config-next: 13.3.1
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

https://gist.github.com/pkellner/72a00cedfc6e2c6a5a90f6ec274454ee

### To Reproduce

when creating project with `npx create-next-app@latest --experimental-app`

replace page.js with

```
import styles from "./page.module.css";

export default function Home() {
  return (
  <div>
    <div className={styles.description}>description</div>
    <div className={styles.largeDescription}>styles.largeDescription</div>
  </div>
  );
}
```

and replace page.module.css with

```
.large-description {
  color: red;
  font-size: 20px;
}

.description {
  color: blue;
  font-size: 30px;
}
```

I expect that second div will be in blue and it appears to not be styled at all.

![image](https://user-images.githubusercontent.com/241170/233892035-86a66acc-6487-4817-b48a-8dd6e836646f.png)

### Describe the Bug

expect that second div will get styled and it does not

### Expected Behavior

expect second div to be styled as blue

### Which browser are you using? (if relevant)

chrome

**Comment by icyJoseph** [comment]
You need to do `styles['large-description ']`

**Comment by pkellner** [comment]
Shouldn't there be an error the way I did it?

**Comment by icyJoseph** [comment]
Nope, and it could get really ugly if you had done something like:

```jsx
    <div className={`${styles.largeDescription}`}>styles.largeDescription</div>
```

Then it'd lay a class name, called undefined... because of type coercion, so luckily in your case, you got nothing.

So it'd render:

```html
    <div class="undefined">styles.largeDescription</div>
```

Which you could make obvious with this bit of CSS,

```css
.undefined {
  background:red;
  width: 10px;
  height: 10px;
  animation: spin 1s infinite linear;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
```

Why would you expect an error at all?

**Comment by pkellner** [comment]
@icyJoseph I had assumed that the error in the attached image was coming from the linter but it seems to be just WebStorm. Not sure why WebStorm thinks so, but I understand what you are saying now.

<img width="1222" alt="image" src="https://user-images.githubusercontent.com/241170/234662912-78f85dd2-5e9c-41e2-ac80-43387a872368.png">

**Comment by jyeharry** [comment]
I think an error/warning is to be expected if `styles.largeDescription` doesn't actually resolve to a defined class. Much like how `styles.largeDescriptionX` threw a warning for @pkellner.

Currently in my setup (a fresh nextjs 13 installation), when I type out a class name it'll autocomplete to the camelcase version of the class name. As @pkellner stated, clicking 'go to declaration' on that autocompleted class name -- in his case, `styles.largeDescription` -- takes me to the correct class in the correct css file. Why then, does `undefined` get returned instead of my modules actual class? Is there a bug in the way NextJS processes css module class names? Or is this a result of my language server coming up with its own mappings for class names which doesn't match NextJS's, therefore requiring one of them (preferably NextJS) to be reconfigured?

**Comment by icyJoseph** [comment]
@jyeharry do you have any extensions/plugins installed that could be culprit? When I do as you describe, I see some typescript compatibility glue file, no css at all.

**Comment by jyeharry** [comment]
@icyJoseph okay I swear when I typed my previous response that it jumped immediately to the css file not the `global.d.ts` file. Now it seems to randomly jump to either `global.d.ts` or my actual css file. So it kinda works.

The only plugins I've installed myself that weren't already installed through `npx create-next-app` are `sass` (I'm actually using scss module files, not css) and `clsx`.

According to this [stack overflow answer](https://stackoverflow.com/questions/35827356/node-react-hyphenated-css-class-names#:~:text=You%20can%20set%20css%20class%20by%20dot%20without%20using%20a%20string%20inside%20a%20bracket) I can configure webpack to map the class names to the format that my autocomplete suggestions are giving me. Though I don't know much about this and therefore whether this is the correct answer. I'll have to try tomorrow.

**Comment by icyJoseph** [comment]
If there's something we could from Next.js' side, to at least print a warning, we probably should. 

- https://github.com/css-modules/css-modules/issues/146

A bit worried about that, this will be fixed soon, in the css-loader, from 2017.

I'll do a bit of documentation/literature reading, and see if this is actionable at all.

**Comment by Surm4** [comment]
> @icyJoseph I had assumed that the error in the attached image was coming from the linter but it seems to be just WebStorm. Not sure why WebStorm thinks so, but I understand what you are saying now.
> 
> And, when I choose "goto declaration" when I leave out the "X" it goes to the module class like I would expect.
> 
> <img alt="image" width="1222" src="https://user-images.githubusercontent.com/241170/234662912-78f85dd2-5e9c-41e2-ac80-43387a872368.png">

yep, facing the same issue in Webstorm 
![Screenshot from 2024-05-13 19-56-12](https://github.com/vercel/next.js/assets/40576160/486a5ac6-b9ab-44d4-9de6-8d9df6130954)
```
 "next": "14.2.3",
```

**Comment by Sam-Kruglov** [comment]
I did some research:
* This is better and gives you a warning if imported class name is undefined but uses Babel instead of [SWC](https://nextjs.org/docs/architecture/nextjs-compiler) https://github.com/gajus/babel-plugin-react-css-modules
* There is a way to add a proxy to get a warning about undefined class like https://github.com/bes/css-module-selector-exists/tree/master or https://github.com/andrewbranch/no-undefined-style-loader/tree/master
* Setting `exportLocalsConvention` to camel case works for me.
* Cleanest approach IMO is to use the default of specifying `styles['my-style']` and have an IDE plugin like https://plugins.jetbrains.com/plugin/9275-react-css-modules highlight in red if the said class doesn't exist. I'm going with this. But if I even get into babel, I'll get that library I mentioned instead

**Comment by elewin** [comment]
It would be nice if the Next team could expose the `exportLocalsConvention` variable in the next config so we don't have to do hacky workarounds to enable it

**Comment by iliyat** [comment]
Any workaround for turbopack in next.js 16?

**Comment by marcoshernanz** [comment]
Thanks for the report. CSS Module class names are exported as written, so `.large-description` should be accessed as `styles['large-description']`. We confirmed this on 13.3.1 and the current canary. Opt-in camel-case exports are being tracked in #91838, with an implementation proposed in #91839, so I’m closing this as expected behavior.
